zilliqa/src/consensus.rs

```diff
@@ -1504,12 +1547,14 @@ impl Consensus {
                 )?;
 
                 // Return whether the head block hash changed as to whether it was new
-                Ok(self.head_block().hash() != current_head.hash())
+                let was_new = self.head_block().hash() != current_head.hash();
+
+                Ok((was_new, response.filter(|_| was_new)))
```

Comment:
Re: my comment about sending stale votes: I still don't think that's sufficient, because `self.head_block()` is only our local view and may be changing all the time while we're syncing. In fact IIRC `was_new` is used to decide whether to request more blocks to continue syncing, so if `was_new == true` it's pretty much an indicator that we're still syncing and therefore still out of date, if I remember the fetching logic correctly.

I still think that unconditionally dropping all `Vote`s (and potentially still broadcasting `Proposal`s just in case) will be worth it, as the very niche edge case where this would cause a `NewView` being far outweighed in my opinion by the upside of significantly cutting down on stale traffic.

Reply:
Right yeah, sorry I misunderstood you originally. I agree.

To restate your argument and confirm my understanding:

If we've requested a pre-existing block and received it, we should assume that we're already too late to vote on that block. In 99% of cases this assumption will be true. In the 1% of cases, where we've requested the most recent block and the next block hasn't been proposed, missing our vote will not bring down the network. In fact, if we wanted to handle this case in the future, I'd suggest we also keep a buffer of the most recent `Proposal` we received, if it didn't form a chain with our existing blocks. However, I don't think that is very useful or worth the effort at the moment.

So I will restore this function back to the previous version :)